cmd/libs/go2idl/lister-gen/generators/lister.go

```diff
@@ -0,0 +1,329 @@
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+
+	groupName := g.groupVersion.Group
+	if g.groupVersion.Group == "core" {
+		groupName = "api"
+	}
+
+	for _, t := range g.types {
+		glog.V(5).Infof("processing type %v", t)
+		m := map[string]interface{}{
+			"group": groupName,
+			"type":  t,
+		}
+
+		namespaced := !extractBoolTagOrDie("nonNamespaced", t.SecondClosestCommentLines)
+		if namespaced {
+			sw.Do(typeListerInterface, m)
+		} else {
+			g.hasNonNamespaced = true
+			sw.Do(typeListerInterface_NonNamespaced, m)
+		}
+
+		sw.Do(typeListerStruct, m)
+		sw.Do(typeListerConstructor, m)
+		sw.Do(typeLister_List, m)
+
+		if namespaced {
+			sw.Do(typeLister_NamespaceLister, m)
+			sw.Do(namespaceListerInterface, m)
+			sw.Do(namespaceListerStruct, m)
+			sw.Do(namespaceLister_List, m)
+			sw.Do(namespaceLister_Get, m)
+		} else {
+			sw.Do(typeLister_NonNamespacedGet, m)
+		}
+	}
+
+	return sw.Error()
+}
+
+var typeListerInterface = `
+// $.type|public$Lister helps list $.type|publicPlural$.
+type $.type|public$Lister interface {
+	// List lists all $.type|publicPlural$ in the indexer.
+	List(selector labels.Selector) (ret []*$.type|raw$, err error)
+	// $.type|publicPlural$ returns an object that can list and get $.type|publicPlural$.
+	$.type|publicPlural$(namespace string) $.type|public$NamespaceLister
+}
+`
+
+var typeListerInterface_NonNamespaced = `
+// $.type|public$Lister helps list $.type|publicPlural$.
+type $.type|public$Lister interface {
+	// List lists all $.type|publicPlural$ in the indexer.
+	List(selector labels.Selector) (ret []*$.type|raw$, err error)
+	// Get retrieves the $.type|public$ from the index for a given name.
+	Get(name string) (*$.type|raw$, error)
+}
+`
+
+var typeListerStruct = `
+// $.type|private$Lister implements the $.type|public$Lister interface.
+type $.type|private$Lister struct {
+	indexer cache.Indexer
+}
+`
+
+var typeListerConstructor = `
+// New$.type|public$Lister returns a new $.type|public$Lister.
+func New$.type|public$Lister(indexer cache.Indexer) $.type|public$Lister {
+	return &$.type|private$Lister{indexer: indexer}
+}
+`
+
+var typeLister_List = `
+// List lists all $.type|publicPlural$ in the indexer.
+func (s *$.type|private$Lister) List(selector labels.Selector) (ret []*$.type|raw$, err error) {
+	err = cache.ListAll(s.indexer, selector, func(m interface{}) {
+		ret = append(ret, m.(*$.type|raw$))
+	})
+	return ret, err
+}
+`
+
+var typeLister_NamespaceLister = `
+// $.type|publicPlural$ returns an object that can list and get $.type|publicPlural$.
+func (s *$.type|private$Lister) $.type|publicPlural$(namespace string) $.type|public$NamespaceLister {
+	return $.type|private$NamespaceLister{indexer: s.indexer, namespace: namespace}
+}
+`
+
+var typeLister_NonNamespacedGet = `
+// Get retrieves the $.type|public$ from the index for a given name.
+func (s *$.type|private$Lister) Get(name string) (*$.type|raw$, error) {
+  key := &$.type|raw${ObjectMeta: api.ObjectMeta{Name: name}}
```

Comment:
I discovered that I can't blindly use `api.ObjectMeta` because the package won't always be `api`. I need to find a way to consistently pick the right package for `ObjectMeta` for each type in question. For right now, it's either `api` or `v1`, but at some point, if we have `v2.ObjectMeta`, I'd love to be able to support that w/o having to modify lister-gen after the fact.